main.py

```diff
@@ -2022,14 +2022,22 @@ def show_hover(self, point, contents):
                 value = item.get("value")
                 language = item.get("language")
             if language:
-                formatted.append("```{}\n{}\n```".format(language, value))
+                formatted.append("```{}\n{}\n```\n".format(language, value))
             else:
-                formatted.append(value)
+                formatted.append("<p class='description'>{}</p>".format(value))
```

Comment:
This line mixes html with markdown and breaks markdown processing.

Reply:
But markdown can have embedded HTML, right?

Reply:
That’s what I thought. Since when exactly can’t you have html in your markdown? Is that an mdpopups thing? 

Also, you don’t have to capitalize “must”, I can read just fine and it just makes you come across as angry and frustrated. If you are, please let me know what I did to have that effect. 

I’ll look into what happens with the python tool tips.

Reply:
Not at all, frustrated. And yes it seems an mdpopups thing to not support embedded html this way as well as forcing `<br>` rather than using paragraphs.

A possible solution was to disable markdown processing `md=False`, create all lines as you did here and use `mdpopups.syntax_highlight()` to get highlighted html for source code embedding it into a `<div class="highlight">` manually. A little more effort was needed to do so, but it's solvable.

But I am concerned about all of this might grow quickly beyond the desired scope as it might create more and more per language edge cases needed to struggle with?

The only thing I am not too fine with is enforcing font-faces on a per package base.

Reply:
> Not at all, frustrated.

Pfieuw 😄 

> A little more effort was needed to do so, but it's solvable.

Yeah, but if the content is in markdown, wouldn't that mean we would also have to convert all that? You use mdpopups in GitGutter right? Is that a custom HTML?

> But I am concerned about all of this might grow quickly beyond the desired scope

It kinda feels like we can't get away with doing nothing (i.e. just push everything through mdpopups) and we definitely don't want do do a lot. I think we can go two ways here:

- get mdpopups / @facelessuser involved to tweak mdpopups based on our findings
- find a middle ground where we process content as little as possible and do as little styling as possible

> The only thing I am not too fine with is enforcing font-faces on a per package base.

Well, it's not a font-face but just a generic family, but I understand your point. I personally see no harm in using `sans-serif` here, but it's not my call.

Reply:
If mdpopups ignores HTML, then that's still okay; it'll be consumed by Sublime's minihtml like we want. If mdpopups doesn't ignore HTML but processes stuff like `*hello*` inside of a HTML tag, then that's a problem. We *could* still work around that, if that so happens the case. @braver, are you willing to experiment wether mdpopups recognizes HTML, but ignores it, *or*, wether mdpopups doesn't recognize HTML and processes it as if it's Markdown?

Reply:
Well, it seems mdpopups will recognise `**bolded**` text (see "function" in one of my earlier screenshots) inside an html element, but it doesn't convert whitespace which messes up the python output.

Also seems like pyls doesn't wrap code snippets in any useful way. That would be another reason to stick with the monospace/editor font at least until servers are more mature and clean up their output.

Reply:
What do you guys need?  You can mix markdown with HTML, but you have to do it right.   There is a TON of info going on here.  Do you guys have  a specific question?

Reply:
@facelessuser 

1.

```
<div>some content</div>
<div>some more content</div>
```

A blank line is rendered between the two divs. It's not padding or margin or can be styled to go away Is that intentional? Is that minihtml or mdpopups?

```
<div>some content</div><div>some more content</div>
```
Has the desired effect, but it's harder to achieve.

2.

I'd expect this markdown:

some text\n
\n
some more text

to be converted to
```
<p>some text</p>
<p>some more text</p>
```

But that doesn't seem to be the case.

Reply:
The main question here is, how can mdpopups convert plain text docstrings which are said to be markdown by language-server-specs to proper looking html, even though each available server returns its own formatted text which is no proper markdown. 😆

Reply:
But seriously, what @braver tried is to create real paragraphs for each returned description item. But somehow mdpopups doesn't create such, but inserts line breaks `<br>` only. Any way to create proper paragraphs?

Reply:
Another question was about preserving white space in normal paragraphs to keep indention of ascii docstrings such as used by python in-tact.

Reply:
Okay. To answer the first question, this is what mdpopups spits out.  What's the problem?

```py
>>> mdpopups.md2html(sublime.active_window().active_view(), 'some text\n\nsome more text\n')
'<p>some text</p><p>some more text</p>'
```

Is that not what you want?

Reply:
Yes it is. For some reason I had the feeling it wouldn't be the case. After some research, the reason is the comments' lines not to be separated with double `\n`. On the other hand, who wants to decide, which lines to separate as paragraph and which one not?

Reply:
What do you mean by preserve whitespace? Can you give me an example?

Reply:
@braver: Paragraphs work with `preserve_whitespace("\n\n".join(formatted))` if we remove `contents = contents.replace('\n\n', '\n&nbsp;\n')` from `preserve_whitespace`.

But this won't solve any problems, because the CSS defines `.mdpopups p { margin: 0.1rem; }` which effectively removes any spacing between paragraphs.

Reply:
If there were a line break between those two paragraphs in your example output, am I correct to assume that would be rendered as a blank line by minihtml as if there was a <br>? That is, minihtml doesn’t eat the white space between two block elements?

Reply:
I'll go ahead and answer some earlier speculation while I'm at it.   Python Markdown (which mdpopups uses) should process markdown in inline tags.  Block tags are excluded.  By default though, I include an extension that will process markdown in block tags if they have the attribute `markdown="1"`.  Usually you to do something like this (notice I have block the content with new lines, also notice that the opening tag is on its own line and the closing on its own):

```
<div markdown="1">

Some **markdown** content.

</div>
```

Markdown doesn't usually care about white space between paragraphs except to separate the paragraphs.  Spaces and tabs usually collapse as they would in any HTML.  Except where syntax specifically cares about whitespace.

Code tags are usually the only element where whitespace is generally perserved.

Reply:
> What do you mean by preserve whitespace? Can you give me an example?

Some (python) - most are actually simple ASCII - docstrings use elements which are indented using several spaces (e.g. Google style docstrings, see my most recent screenshot). Passing those strings to markdown would squash down all the spaces to one. IMHO, this is default markdown behavior, but doesn't work well, when trying to handle ASCII as markdown.

Reply:
So, if you want to protect HTML, put it in a block `div` and you are okay.  Nothing **should** get parsed.

But there is always a caveat. Python Markdown has a weakness in the way it processes fenced blocks.  I won't go into elaborate details, but fenced blocks are parsed as a pre processor, then the file is parsed for HTML blocks and sometimes a fenced block in an HTML block that is not meant to get parsed turns into soemthing like this:

before:

````
<div>

```
somecode
```
 
</div>

````

after:

````
<div>

w23134870280

</div>
````

The weird letters and numbers would be a place holder that never gets restored.

Basically fences and HTML are handled as separate  tasks when really you kind of need something more advanced that handles them simultaneously (or inline together).

Reply:
I guess you could turn off fences and not have to deal with that.

Reply:
I should also mention that setting that setting `md=False` when calling show popup will just pass through HTML without applying markdown.

Anyways, if you are hoping to have docstrings parsed as markdown-ish, you may have to process yourselves as markdown-ish may not work.  It's possible you could write your own extensions for Python Markdown and use those though....

Anyways, it sounds like you guys have a lot to figure out.  If you have more specific questions, I'd be happy to answer them.

Reply:
Maybe we should handle this server-side?

Reply:
@rwols in what sense exactly? Getting the servers to standardize their output?

Reply:
Learned *so much* in the above. Adding the `markdown='1'` attribute allows us to wrap the content in something without losing the markdown processing. Also, minihtml isn't as weird as I feared it might be, but I *am* getting a <br> where I don't want one. 

I'll fiddle with it some more.

```
<p><div class="highlight"><pre>function&nbsp;scaleOrdinal&lt;string&gt;(range?:&nbsp;string[]):&nbsp;ScaleOrdinal&lt;string,&nbsp;string&gt;&nbsp;(+1&nbsp;overload)<br></pre></div><br /><div class='description' markdown='1'><strong>function</strong> <em>(exported, ambient)</em></div><div class='description' markdown='1'>Constructs a new ordinal scale with an empty domain and the specified range.<br />If a range is not specified, it defaults to the empty array; an ordinal scale always returns undefined until a non-empty range is defined.<br /><br />By default, the domain is configured to generate implicitly, if the scale is invoked with an unknown value.<br />See the "unknown(...)" method of the scale to change this behavior.<br /><br />The generic corresponds to the data type of range elements.</div></p>
```

Reply:
>  in what sense exactly? Getting the servers to standardize their output?

Yes. Different languages have different traditions for documentation. e.g. Python has its own ascii style, C++ has Doxygen, Java has JavaDoc, etc.